File: ui/src/api/rules.ts
import { api } from "./client";

export interface RuleInfo {
  id: string;
  name: string;
  description: string;
  version: number;
  enabled: boolean;
  severity: string;
  collector: string;
  targets: string[];
  matcher_type: string;
  owasp: string[];
  tags: string[];
  source: string;
  test_count: number;
}

export interface MatcherSpec {
  Type: string;
  Pattern?: string;
  Keywords?: string[];
  Prefixes?: string[];
  CaseInsensitive?: boolean;
  MatchMode?: string;
  Operator?: string;
  Matchers?: MatcherSpec[];
  Charset?: string;
  Threshold?: number;
  MinLength?: number;
}

export interface TestCase {
  Input: string;
  ShouldMatch: boolean;
  Description: string;
}

export interface RuleDetail extends RuleInfo {
  matcher: MatcherSpec;
  tests: TestCase[];
}

export async function fetchRules(params?: {
  collector?: string;
  severity?: string;
  tag?: string;
}): Promise<{ rules: RuleInfo[]; total: number }> {
  const searchParams: Record<string, string> = {};
  if (params?.collector) searchParams["collector"] = params.collector;
  if (params?.severity) searchParams["severity"] = params.severity;
  if (params?.tag) searchParams["tag"] = params.tag;
  return api
    .get("rules", { searchParams })
    .json<{ rules: RuleInfo[]; total: number }>();
}

export async function fetchRuleDetail(id: string): Promise<RuleDetail> {
  return api.get(`rules/${encodeURIComponent(id)}`).json<RuleDetail>();
}

File: server/internal/api/handlers/rules.go
package handlers

import (
	"net/http"
	"strings"

	"github.com/adithyan-ak/agenthound/sdk/rules"
	"github.com/go-chi/chi/v5"
)

type RulesHandler struct {
	engine *rules.Engine
}

func NewRulesHandler(engine *rules.Engine) *RulesHandler {
	return &RulesHandler{engine: engine}
}

type RuleResponse struct {
	ID          string   `json:"id"`
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Version     int      `json:"version"`
	Enabled     bool     `json:"enabled"`
	Severity    string   `json:"severity"`
	Collector   string   `json:"collector"`
	Targets     []string `json:"targets"`
	MatcherType string   `json:"matcher_type"`
	OWASP       []string `json:"owasp,omitempty"`
	Tags        []string `json:"tags"`
	Source      string   `json:"source"`
	TestCount   int      `json:"test_count"`
}

type RuleDetailResponse struct {
	RuleResponse
	Matcher rules.MatcherSpec `json:"matcher"`
	Tests   []rules.TestCase  `json:"tests"`
}

func ruleToResponse(r rules.Rule) RuleResponse {
	tags := r.Tags
	if tags == nil {
		tags = []string{}
	}
	targets := r.Scope.Targets
	if targets == nil {
		targets = []string{}
	}
	return RuleResponse{
		ID:          r.ID,
		Name:        r.Name,
		Description: r.Description,
		Version:     r.Version,
		Enabled:     true,
		Severity:    r.Severity,
		Collector:   r.Scope.Collector,
		Targets:     targets,
		MatcherType: r.Matcher.Type,
		OWASP:       r.OWASP,
		Tags:        tags,
		Source:      r.Source,
		TestCount:   len(r.Tests),
	}
}

func (h *RulesHandler) HandleList(w http.ResponseWriter, r *http.Request) {
	if h.engine == nil {
		WriteJSON(w, http.StatusOK, map[string]any{
			"rules": []RuleResponse{},
			"total": 0,
		})
		return
	}

	collector := r.URL.Query().Get("collector")
	severity := r.URL.Query().Get("severity")
	tag := r.URL.Query().Get("tag")

	all := h.engine.Rules()
	var filtered []RuleResponse
	for _, rule := range all {
		if collector != "" && !strings.EqualFold(rule.Scope.Collector, collector) {
			continue
		}
		if severity != "" && !strings.EqualFold(rule.Severity, severity) {
			continue
		}
		if tag != "" && !hasTag(rule.Tags, tag) {
			continue
		}
		filtered = append(filtered, ruleToResponse(rule))
	}
	if filtered == nil {
		filtered = []RuleResponse{}
	}

	WriteJSON(w, http.StatusOK, map[string]any{
		"rules": filtered,
		"total": len(filtered),
	})
}

func (h *RulesHandler) HandleGet(w http.ResponseWriter, r *http.Request) {
	if h.engine == nil {
		WriteNotFound(w, "rule not found")
		return
	}

	id := chi.URLParam(r, "id")
	for _, rule := range h.engine.Rules() {
		if rule.ID == id {
			tests := rule.Tests
			if tests == nil {
				tests = []rules.TestCase{}
			}
			WriteJSON(w, http.StatusOK, RuleDetailResponse{
				RuleResponse: ruleToResponse(rule),
				Matcher:      rule.Matcher,
				Tests:        tests,
			})
			return
		}
	}

	WriteNotFound(w, "rule not found")
}

func hasTag(tags []string, target string) bool {
	for _, t := range tags {
		if strings.EqualFold(t, target) {
			return true
		}
	}
	return false
}
